### PR TITLE
V3: engine work

### PR DESCRIPTION
**Pull request by sougou**
@alainjobart 
This is the migration of the primitives into their own package.

* Pulling out the routing operations from the router was too tricky. So, I've left that part alone for now. I'm even wondering if it may be unnatural to pull that out. The IN clause function that converts input column values to per-shard maps is particularly hairy.
* Gateway didn't have the right level of abstraction. It was too low level, underneath ScatterConn.
* Looks like Router is actually using V1 (ScatterConn), not V2 (Resolver). But Resolver doesn't have the equivalent of ExecuteMulti. It has ExecuteEntityId, which is inconvenient for V3. 
* I considered going partial V1/V2, but I think that will cause more confusion. Eventually, I think we should look at extending Resolver to support everything that V3 needs.
* I ran into circular dependencies that got resolved by moving VSchema and Vindex into vindexes.
* Renamed a bunch of things and updated comments.

---
This change is [Reviewable](https://reviewable.io/reviews/youtube/vitess/1601)

### CODE REVIEW

File: go/vt/vtgate/engine/primitive.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+// ListVarName is a reserved bind var name for list vars.
+// This is used for sending different IN clause values
+// to different shards.
+const ListVarName = "__vals"
+
+// VCursor defines the interface the engine will use
+// to execute routes.
+type VCursor interface {
+	ExecRoute(route *Route, joinvars map[string]interface{}) (*sqltypes.Result, error)
```
Comment (alainjobart) [inline go/vt/vtgate/engine/primitive.go]: This should be ExecuteRoute, next one StreamExecuteRoute.

File: go/vt/vtgate/engine/primitive.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+// VTGate needs to maintain a cache of plans. It uses LRUCache, which
+// in turn requires its objects to define a Size function.
+func (pln *Plan) Size() int {
+	return 1
+}
+
+// Primitive is the interface that needs to be satisfied by
+// all primitives of a plan.
+type Primitive interface {
+	Exec(vcursor VCursor, joinvars map[string]interface{}, wantields bool) (*sqltypes.Result, error)
```
Comment (alainjobart) [inline go/vt/vtgate/engine/primitive.go]: This should be Execute, next StreamExecute, to be consistent with the rest of the interfaces all over the place.

File: go/vt/vtgate/router.go [hunk 3]
```diff
@@ -68,135 +68,50 @@ func (rtr *Router) Execute(ctx context.Context, sql string, bindVars map[string]
...
-// execRoute executes the route query for all route opcodes.
-func (rtr *Router) execRoute(vcursor *requestContext, route *planbuilder.Route) (*sqltypes.Result, error) {
+// ExecRoute executes the route query for all route opcodes.
+func (rtr *Router) ExecRoute(vcursor *requestContext, route *engine.Route, joinvars map[string]interface{}) (*sqltypes.Result, error) {
 	saved := copyBindVars(vcursor.bindVars)
 	defer func() { vcursor.bindVars = saved }()
-	for k := range route.JoinVars {
-		vcursor.bindVars[k] = vcursor.JoinVars[k]
+	for k := range joinvars {
+		vcursor.bindVars[k] = joinvars[k]
```
Comment (alainjobart) [inline go/vt/vtgate/router.go]: Nit here:, would this be more efficient:

for k, v := range joinvars {
  vcursor.bindVars[k] = v
}

?

**Comment by alainjobart** [comment]
LGTM overall, couple comments.

**Comment by sougou** [comment]
All comments addressed.

**Pull request merged**